File: src/restricted-extras-qscintilla/libqscicommandset.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const qscicommand_enums = @import("libqscicommand.zig").enums;
const std = @import("std");

/// https://www.riverbankcomputing.com/static/Docs/QScintilla/classQsciCommandSet.html
pub const qscicommandset = struct {
    /// [Qt documentation](https://www.riverbankcomputing.com/static/Docs/QScintilla/classQsciCommandSet.html)
    ///
    /// ``` self: QtC.QsciCommandSet, qs: QtC.QSettings ```
    pub fn ReadSettings(self: ?*anyopaque, qs: ?*anyopaque) bool {
        return qtc.QsciCommandSet_ReadSettings(@ptrCast(self), @ptrCast(qs));
    }

    /// [Qt documentation](https://www.riverbankcomputing.com/static/Docs/QScintilla/classQsciCommandSet.html)
    ///
    /// ``` self: QtC.QsciCommandSet, qs: QtC.QSettings ```
    pub fn WriteSettings(self: ?*anyopaque, qs: ?*anyopaque) bool {
        return qtc.QsciCommandSet_WriteSettings(@ptrCast(self), @ptrCast(qs));
    }

    /// [Qt documentation](https://www.riverbankcomputing.com/static/Docs/QScintilla/classQsciCommandSet.html)
    ///
    /// ``` self: QtC.QsciCommandSet, allocator: std.mem.Allocator ```
    pub fn Commands(self: ?*anyopaque, allocator: std.mem.Allocator) []QtC.QsciCommand {
        const _arr: qtc.libqt_list = qtc.QsciCommandSet_Commands(@ptrCast(self));
        defer qtc.libqt_free(_arr.data);
        const _ret = allocator.alloc(QtC.QsciCommand, _arr.len) catch @panic("qscicommandset.Commands: Memory allocation failed");
        const _data: [*]QtC.QsciCommand = @ptrCast(@alignCast(_arr.data));
        @memcpy(_ret, _data[0.._arr.len]);
        return _ret;
    }

    /// [Qt documentation](https://www.riverbankcomputing.com/static/Docs/QScintilla/classQsciCommandSet.html)
    ///
    /// ``` self: QtC.QsciCommandSet ```
    pub fn ClearKeys(self: ?*anyopaque) void {
        qtc.QsciCommandSet_ClearKeys(@ptrCast(self));
    }

    /// [Qt documentation](https://www.riverbankcomputing.com/static/Docs/QScintilla/classQsciCommandSet.html)
    ///
    /// ``` self: QtC.QsciCommandSet ```
    pub fn ClearAlternateKeys(self: ?*anyopaque) void {
        qtc.QsciCommandSet_ClearAlternateKeys(@ptrCast(self));
    }

    /// [Qt documentation](https://www.riverbankcomputing.com/static/Docs/QScintilla/classQsciCommandSet.html)
    ///
    /// ``` self: QtC.QsciCommandSet, key: i32 ```
    pub fn BoundTo(self: ?*anyopaque, key: i32) QtC.QsciCommand {
        return qtc.QsciCommandSet_BoundTo(@ptrCast(self), @intCast(key));
    }

    /// [Qt documentation](https://www.riverbankcomputing.com/static/Docs/QScintilla/classQsciCommandSet.html)
    ///
    /// ``` self: QtC.QsciCommandSet, command: qscicommand_enums.Command ```
    pub fn Find(self: ?*anyopaque, command: i64) QtC.QsciCommand {
        return qtc.QsciCommandSet_Find(@ptrCast(self), @intCast(command));
    }

    /// [Qt documentation](https://www.riverbankcomputing.com/static/Docs/QScintilla/classQsciCommandSet.html)
    ///
    /// ``` self: QtC.QsciCommandSet, qs: QtC.QSettings, prefix: []const u8 ```
    pub fn ReadSettings2(self: ?*anyopaque, qs: ?*anyopaque, prefix: []const u8) bool {
        const prefix_Cstring = prefix.ptr;
        return qtc.QsciCommandSet_ReadSettings2(@ptrCast(self), @ptrCast(qs), prefix_Cstring);
    }

    /// [Qt documentation](https://www.riverbankcomputing.com/static/Docs/QScintilla/classQsciCommandSet.html)
    ///
    /// ``` self: QtC.QsciCommandSet, qs: QtC.QSettings, prefix: []const u8 ```
    pub fn WriteSettings2(self: ?*anyopaque, qs: ?*anyopaque, prefix: []const u8) bool {
        const prefix_Cstring = prefix.ptr;
        return qtc.QsciCommandSet_WriteSettings2(@ptrCast(self), @ptrCast(qs), prefix_Cstring);
    }
};

File: src_original/restricted-extras-qscintilla/libqscicommandset.cpp
#include <QList>
#include <QSettings>
#include <qscicommandset.h>
#include "libqscicommandset.h"
#include "libqscicommandset.hxx"

bool QsciCommandSet_ReadSettings(QsciCommandSet* self, QSettings* qs) {
    return self->readSettings(*qs);
}

bool QsciCommandSet_WriteSettings(QsciCommandSet* self, QSettings* qs) {
    return self->writeSettings(*qs);
}

libqt_list /* of QsciCommand* */ QsciCommandSet_Commands(QsciCommandSet* self) {
    QList<QsciCommand*>& _ret = self->commands();
    // Convert QList<> from C++ memory to manually-managed C memory
    QsciCommand** _arr = static_cast<QsciCommand**>(malloc(sizeof(QsciCommand*) * (_ret.size() + 1)));
    for (qsizetype i = 0; i < _ret.size(); ++i) {
        _arr[i] = _ret[i];
    }
    libqt_list _out;
    _out.len = _ret.size();
    _out.data = static_cast<void*>(_arr);
    return _out;
}

void QsciCommandSet_ClearKeys(QsciCommandSet* self) {
    self->clearKeys();
}

void QsciCommandSet_ClearAlternateKeys(QsciCommandSet* self) {
    self->clearAlternateKeys();
}

QsciCommand* QsciCommandSet_BoundTo(const QsciCommandSet* self, int key) {
    return self->boundTo(static_cast<int>(key));
}

QsciCommand* QsciCommandSet_Find(const QsciCommandSet* self, int command) {
    return self->find(static_cast<QsciCommand::Command>(command));
}

bool QsciCommandSet_ReadSettings2(QsciCommandSet* self, QSettings* qs, const char* prefix) {
    return self->readSettings(*qs, prefix);
}

bool QsciCommandSet_WriteSettings2(QsciCommandSet* self, QSettings* qs, const char* prefix) {
    return self->writeSettings(*qs, prefix);
}
